internal/pkg/grpc/tls/cert.go

```diff
@@ -0,0 +1,185 @@
+		f(cert)
+	}
+	return nil
+}
+
+type userDataCertificateProvider struct {
+	data *userdata.OSSecurity
+}
+
+func (p *userDataCertificateProvider) GetCertificate(h *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	cert, err := tls.X509KeyPair(p.data.Identity.Crt, p.data.Identity.Key)
+	return &cert, err
+}
+
+func (p *userDataCertificateProvider) UpdateCertificate(h *tls.ClientHelloInfo, cert *tls.Certificate) error {
+	// No-op
+	return nil
+}
+
+type renewingFileCertificateProvider struct {
+	singleCertificateProvider
+
+	certFile string
+
+	// For now, this is using the complete userdata object.  It should probably
+	// be pared down to just what is necessary, later.  However, the current TLS
+	// generator code requires the whole thing... so we do, too.
+	data *userdata.UserData
+
+	g *gen.Generator
+}
+
+// NewRenewingFileCertificateProvider returns a new CertificateProvider which
+// manages and updates its certificates from trustd, storing a cache file copy.
+//
+// TODO: the flow here is a bit wonky, but it should be fixable after we change
+// to have the default be ephemeral node certificates.  Until then, however, we
+// are doing a dance between the userdata-stored cert, the filesystem-cached
+// cert, and the memory-cached cert.
+func NewRenewingFileCertificateProvider(ctx context.Context, data *userdata.UserData) (CertificateProvider, error) {
+	g, err := gen.NewGenerator(data, constants.TrustdPort)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create TLS generator")
+	}
+
+	p := &renewingFileCertificateProvider{
+		g:        g,
+		data:     data,
+		certFile: constants.NodeCertFile,
+	}
+
+	if err = p.loadInitialCert(); err != nil {
+		return nil, errors.Wrap(err, "failed to load initial certificate")
+	}
+
+	go p.manageUpdates(ctx)
+
+	return p, nil
+}
+
+func (p *renewingFileCertificateProvider) loadInitialCert() error {
+
+	// TODO: eventually, we will reverse this priority, and have the override
+	// come from the userdata.  For now, however, we use the local file to
+	// override the userdata, because we _always_ have userdata certs, and the
+	// userdata is intended to be immutable.
+
+	data, err := ioutil.ReadFile(p.certFile)
+	if err != nil || len(data) == 0 {
+		// If we cannot read the cert from the file, then we use the userdata-supplied one
+		data = p.data.Security.OS.Identity.Crt
+	}
+
+	cert, err := tls.X509KeyPair(data, p.data.Security.OS.Identity.Key)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse cert and key into a TLS Certificate")
+	}
+
+	return p.UpdateCertificate(nil, &cert)
+}
+
+func (p *renewingFileCertificateProvider) manageUpdates(ctx context.Context) {
+	var nextRenewal = constants.NodeCertRenewalInterval
+
+	for ctx.Err() == nil {
+		if c, _ := p.GetCertificate(nil); c != nil { // nolint: errcheck
+			if len(c.Certificate) > 0 {
+				cert, err := x509.ParseCertificate(c.Certificate[0])
+				if err == nil {
+					nextRenewal = time.Until(cert.NotAfter) / 2
```

Comment:
Ahh I think I understand. We default to a minute, but then transition to half the expiry date.

Reply:
Yes, but even then, 1 minute is too aggressive.  That was just for testing.